### PR TITLE
Respect ReportSuppressedDiagnostics option for DiagnosticSuppressor

### LINKED ISSUE #41584: CompilationWithAnalyzers.GetAllDiagnosticsAsync() does not respect CompilationWithAnalyzersOptions.ReportSuppressedDiagnostics
**Version Used**: 3.3.1

**Steps to Reproduce**:

1. Set `CompilationOptions.ReportSuppressedDiagnostics` to false
2. Get a `CompilationWithAnalyzers`
3. Call `GetAllDiagnosticsAsync()`

**Expected Behavior**:

The suppressed diagnostics are not included in the result.

**Actual Behavior**:

The suppressed diagnostics are included in the result.

### PR DESCRIPTION
**Pull request by safu9**
This PR fixes #41584.

Diagnostics suppressed programatically by DiagnosticSuppressor were not excluded properly when ReportSuppressedDiagnostics option is set to true.

### CODE REVIEW

**Comment by safu9** [comment]
@sharwell @mavasani Could you review this PR?

**Review by mavasani — CHANGES_REQUESTED** [review]
This compilation option is not to determine whether or not diagnostic suppressors should run. The option indicates whether or suppressed diagnostics should be filtered out from final diagnostics output. Basically, this option is unrelated to suppressors.

**Comment by safu9** [comment]
I mentioned this in #52953.
This option works properly with many ways to suppress diagnostics such as `pragma warning disable` and `SuppressMessageAttribute`.
I think the only exception is DiagnosticSuppressor.
This option does not exclude diagnostics suppressed by DiagnosticSuppressor for now.

The current program executes following steps if this option is true.
1. List up all diagnostics and call this method. https://github.com/dotnet/roslyn/blob/2f20729b0b4bc78865fcc663aa4cf33323b723d3/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs#L1070
2. At line 1072, `FilterDiagnosticsSuppressedInSource` exclude suppressed diagnostics in source, which is fine.
3. At line 1073, `ApplyProgrammaticSuppressions` check whether each diagnostic is suppressed by DiagnosticSuppressor.

In 3rd step, `ApplyProgrammaticSuppressions` is expected exclude suppressed diagnostics, but they actually only set `Diagnostic.IsSuppressed` to true and return them.
I fixed this behaviour of `ApplyProgrammaticSuppressions` in this PR.
This PR doesn't prevent DiagnosticSuppressor from runnning.

**Comment by Youssef1313** [comment]
What I understand from the PR is that currently, ReportSuppressedDiagnostics isn't respected for DiagnosticSuppressors. However, I see DiagnosticSuppressor tests in roslyn-analyzers relying on ReportSuppressedDiagnostics property.

https://github.com/dotnet/roslyn-analyzers/blob/a534bf8c8310f375d163f701ade001e47ef55fb2/src/Roslyn.Diagnostics.Analyzers/UnitTests/RelaxTestNamingSuppressorTests.cs#L24

Not sure what's going on?

**Review by sharwell — APPROVED** [review]

**Comment by sharwell** [comment]
I agree with the analysis in https://github.com/dotnet/roslyn/pull/53045#issuecomment-832579578. This oversight is the main reason we don't have any support for testing diagnostic suppressors in Microsoft.CodeAnalysis.Testing.

**Comment by Youssef1313** [comment]
> I agree with the analysis in [#53045 (comment)](https://github.com/dotnet/roslyn/pull/53045#issuecomment-832579578). This oversight is the main reason we don't have any support for testing diagnostic suppressors in Microsoft.CodeAnalysis.Testing.

FYI - I managed to write tests for suppressors before. https://github.com/dotnet/efcore/pull/24151/ (might not be the best way - but you may be able to use a similar idea for roslyn-sdk until this PR is merged and shipped)

**Comment by mavasani** [comment]
The implementation in this PR is wrong, the filtering is done at the wrong place. Note that without your change, all programmatically suppressed diagnostics from suppressors will always go through the below code path and info diagnostics will be reported in binary logs for them, regardless of the value of this option: https://github.com/dotnet/roslyn/blob/a6b8bb581cac6b1016a45d1a50d96b60a0d584f1/src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs#L560-L571

With your change, this will no longer be true when `ReportSuppressedDiagnostics` option is false. Note that **always** reporting these info diagnostics corresponding to diagnostics suppressed by suppressors was a **hard** requirement in the original design approval of suppressors. Without these info diagnostics in binlogs, end users would have no idea what caused their diagnostics to be suppressed.

The correct fix here would be to revert your changes and move out the filtering to be done just after the callsite inside `CompilationWithAnalyzers` to apply programmatic suppressions: https://github.com/dotnet/roslyn/blob/a6b8bb581cac6b1016a45d1a50d96b60a0d584f1/src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs#L468

Also, please add unit tests for this change.

**Comment by Youssef1313** [comment]
@mavasani Wouldn't that make the method `FilterDiagnosticsSuppressedInSourceOrByAnalyzers` misleading? It won't be filtering anything by `DiagnosticSuppressor`s as the name suggests?

**Comment by mavasani** [comment]
@Youssef1313 Naming is a minor thing for a private method. We cannot cause a functionality break for the most important requirement in the DiagnosticSuppressor design.

**Comment by Youssef1313** [comment]
@mavasani Yeah certainly. I meant that we could rename it as a minor cleanup.

**Comment by safu9** [comment]
@mavasani Thank you for the review. I'll fix the implementation of `CompilationWithAnalyzers`.

**Comment by safu9** [comment]
@mavasani I have rewrited the implementation and add some tests.
This should not change behaviour of `CommonCompiler`, and just fix `CompilationWithAnalyzers`.
`CommonCompiler` is using `AnalyzerDriver.ApplyProgramaticSuppressions`, so I revert its changes and added a new method to filter suppressed diagnostics.

**Comment by safu9** [comment]
@mavasani Could you re-review this?

**Comment by Youssef1313** [comment]
@mavasani Can you take a look please? Thanks!

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 1]
```diff
@@ -1101,6 +1101,31 @@ private ImmutableArray<Diagnostic> FilterDiagnosticsSuppressedInSourceOrByAnalyz
...
+                if (!reportSuppressedDiagnostics && diagnostic.IsSuppressed)
+                {
+                    // Diagnostic suppressed by analyzer.
+                    continue;
+                }
+
+                builder.Add(diagnostic);
+            }
+
+            return builder.ToImmutable();
```
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:1126]:
```suggestion
            if (compilation.Options.ReportSuppressedDiagnostics || diagnostics.All(d => !d.IsSuppressed))
               return diagnostics;

            return diagnostics.WhereAsArray(d => !d.IsSuppressed);
```

File: src/Compilers/Test/Core/Diagnostics/DiagnosticExtensions.cs [hunk 2]
```diff
@@ -273,6 +273,21 @@ public static TCompilation VerifyDiagnostics<TCompilation>(this TCompilation c,
...
+            DiagnosticAnalyzer[] analyzers,
+            AnalyzerOptions options = null,
+            Action<Exception, DiagnosticAnalyzer, Diagnostic> onAnalyzerException = null,
+            CancellationToken cancellationToken = default)
+            where TCompilation : Compilation
+        {
+            // Verify suppressed diagnostics are filtered when reportSuppressedDiagnostics is false.
+            c = c.GetCompilationWithAnalyzerDiagnostics(analyzers, options, onAnalyzerException, reportSuppressedDiagnostics: false, includeCompilerDiagnostics: true, cancellationToken, out var diagnostics);
+            diagnostics = diagnostics.WhereAsArray(d => d.IsSuppressed);
+            Assert.Empty(diagnostics);
```
Comment (mavasani) [inline src/Compilers/Test/Core/Diagnostics/DiagnosticExtensions.cs:287]: This seems to be a redundant check. Code below from [line 319-322](https://github.com/dotnet/roslyn/pull/53045/files#diff-a0b279d58233deddedac1fca458609ef81a83be228dee99e39476ec043b5ff87R319-R322) already does this validation.
```suggestion
```

Comment (mavasani) [reply src/Compilers/Test/Core/Diagnostics/DiagnosticExtensions.cs:287]: I mean the below existing code in `GetCompilationWithAnalyzerDiagnostics`:

```cs
            if (!reportSuppressedDiagnostics)
            {
                Assert.True(diagnostics.All(d => !d.IsSuppressed));
            }
```

**Review by mavasani — APPROVED** [review]
Looks good to me. I have couple of suggestions to further simplify the code changes.

**Comment by mavasani** [comment]
@dotnet/roslyn-compiler for reviews

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 3]
```diff
@@ -1101,6 +1101,22 @@ private ImmutableArray<Diagnostic> FilterDiagnosticsSuppressedInSourceOrByAnalyz
...
 
+        internal ImmutableArray<Diagnostic> ApplyProgrammaticSuppressionsAndFilterDiagnostics(ImmutableArray<Diagnostic> reportedDiagnostics, Compilation compilation)
+        {
+            if (reportedDiagnostics.IsEmpty)
+            {
+                return reportedDiagnostics;
+            }
+
+            var diagnostics = ApplyProgrammaticSuppressions(reportedDiagnostics, compilation);
+            if (compilation.Options.ReportSuppressedDiagnostics || diagnostics.All(d => !d.IsSuppressed))
```
Comment (cston) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:1112]: >|| diagnostics.All(d => !d.IsSuppressed) [](http://example.com/codeflow?start=64&length=40)

Is this check necessary? (`WhereAsArray()` below will return the original array without allocations if all items satisfy the predicate.)

File: src/Compilers/Test/Core/Diagnostics/DiagnosticExtensions.cs [hunk 4]
```diff
@@ -273,6 +273,19 @@ public static TCompilation VerifyDiagnostics<TCompilation>(this TCompilation c,
...
 
+        public static TCompilation VerifySuppressedAndFilteredDiagnostics<TCompilation>(
+            this TCompilation c,
+            DiagnosticAnalyzer[] analyzers,
+            AnalyzerOptions options = null,
+            Action<Exception, DiagnosticAnalyzer, Diagnostic> onAnalyzerException = null,
+            CancellationToken cancellationToken = default)
+            where TCompilation : Compilation
+        {
+            // Verify suppressed diagnostics are filtered when reportSuppressedDiagnostics is false.
```
[the remark this answers is not in the document]
Comment (safu9) [reply src/Compilers/Test/Core/Diagnostics/DiagnosticExtensions.cs:284]: Please see this comment. https://github.com/dotnet/roslyn/pull/53045#discussion_r790530597
`GetCompilationWithAnalyzerDiagnostics` already contains the verification.

[the remark this answers is not in the document]
Comment (cston) [reply src/Compilers/Test/Core/Diagnostics/DiagnosticExtensions.cs:284]: > `GetCompilationWithAnalyzerDiagnostics` already contains the verification.

Thanks. Please consider adding a comment here that it's handled in `GetCompilationWithAnalyzerDiagnostics`.

**Review by cston — APPROVED** [review]

**Comment by mavasani** [comment]
@dotnet/roslyn-compiler for second review.

**Review by 333fred — APPROVED** [review]

**Pull request merged**